mono/utils/mono-cgroup.c

```diff
@@ -0,0 +1,958 @@
+		goto done;
+	}
+
+	// "max" means no cpu limit
+	if (strcmp("max", max_quota_string) == 0)
+		goto done;
+
+	errno = 0;
+	quota = strtoll(max_quota_string, &endptr, 10);
+	if (max_quota_string == endptr || errno != 0)
+		goto done;
+
+	period = strtoll(period_string, &endptr, 10);
+	if (period_string == endptr || errno != 0)
+		goto done;
+
+	computeCpuLimit(period, quota, val);
+	result = TRUE;
+
+done:
+	if (file)
+		fclose(file);
+	free(filename);
+	free(line);
+
+	return result;
+}
+
+/**
+ *
+ * @brief Compute the CPU limit based on the CGroup data
+ *
+ * @param[in] period
+ * @param[in] quota - Limit found in sysfs
+ * @param[out] Number of CPU usable
+ *
+ */
+static void 
+computeCpuLimit(long long period, long long quota, uint32_t *val)
+{
+	// Cannot have less than 1 CPU
+	if (quota <= period) {
+		*val = 1;
+		return;
+	}
+
+	// Calculate cpu count based on quota and round it up
+	double cpu_count = (double) quota / period  + 0.999999999;
+	*val = (cpu_count < UINT32_MAX) ? (uint32_t)cpu_count : UINT32_MAX;
+}
+
+/**
+ *
+ * @brief Read the CGroup CPU data from sysfs
+ *
+ * @param[in] subsystemFileName - sysfs File containing data
+ * @returns CPU CGroup value
+ *
+ */
+static long long 
+readCpuCGroupValue(const char *subsystemFilename)
+{
+	char *filename = NULL;
+	gboolean result = FALSE;
+	long long val;
+
+	if (s_cpu_cgroup_path == NULL)
+		return -1;
+
+	if (asprintf(&filename, "%s%s", s_cpu_cgroup_path, subsystemFilename) < 0)
+		return -1;
+
+	result = readLongLongValueFromFile(filename, &val);
+	free(filename);
+	if (!result)
+		return -1;
+
+	return val;
+}
+
+/**
+ *
+ * @brief Read a long long value from a file
+ *
+ * @param[in] fileName - sysfs File containing data
+ * @param[out] val - Value read
+ * @returns Success indicator
+ *
+ */
+static gboolean 
+readLongLongValueFromFile(const char *filename, long long *val)
+{
+	gboolean result = FALSE;
+	char *line = NULL;
+	size_t lineLen = 0;
+	char *endptr = NULL;
+
+	if (val == NULL)
+		return FALSE;
+
+	FILE *file = fopen(filename, "r");
+	if (file == NULL)
+		return FALSE;
+
+	if (getline(&line, &lineLen, file) != -1) {
+		errno = 0;
+		*val = strtoll(line, &endptr, 10);
+		if (line != endptr && errno == 0)
+			result = TRUE;
+	}
+
+	fclose(file);
+	free(line);
+	return result;
+}
+
+/**
+ *
+ * @brief Interrogate the cgroup CPU values to determine if there's
+ * a limit on CPUs
+ *
+ * @param[out] val - pointer to the result area
+ * @returns True or False depending if a limit was found
+ *
+ * Interrogate the cgroup CPU values to determine if there's
+ * a limit on CPUs
+ */
+gboolean 
+getCpuLimit(guint *val)
+{
+	if (s_mem_stat_n_keys == 0)
+		initialize();
+
+	if (s_cgroup_version == 0)
+		return FALSE;
+	else if (s_cgroup_version == 1)
+		return getCGroup1CpuLimit(val);
+	else if (s_cgroup_version == 2)
+		return getCGroup2CpuLimit(val);
+	else {
+		g_assert(!"Unknown cgroup version.");
```

Comment:
```suggestion
		g_error("Unknown cgroup version.");
```

Reply:
we should also change the other asserts like this in the file.

Reply:
Changed all `g_assert(!` to `g_error(`